Adding app

jkarimi91 opened the pull request.

jay-karimi committed:
Adding app

bvallelunga commented on app/encoder.py:
```diff
@@ -0,0 +1,223 @@
+    return params[nloaded - 1]
+
+
+def embd(X, ndim, scope='embedding'):
```
should that be spelled `embed`?

bvallelunga replied:
should that be spelled` embed`?

bvallelunga commented on app/encoder.py:
```diff
@@ -0,0 +1,223 @@
+        return h
+
+
+def fc(x, nout, act, wn=False, bias=True, scope='fc'):
```
can we have more descriptive function names?

bvallelunga replied:
Can you add more documentation to what these methods do?

bvallelunga commented on app/encoder.py:
```diff
@@ -0,0 +1,223 @@
+        return h
+
+
+def mlstm(inputs, c, h, M, ndim, scope='lstm', wn=False):
```
Same here

jkarimi91 replied:
mlstm is the name of the underlying recurrent neural network model; it's a multiplicative lstm. I will add a docstring

bvallelunga commented on app/encoder.py:
```diff
@@ -0,0 +1,223 @@
+            gmx = tf.get_variable("gmx", [ndim], initializer=load_params)
+            gmh = tf.get_variable("gmh", [ndim], initializer=load_params)
+
+    if wn:
```
what is wn?

jkarimi91 replied:
after skimming the paper, it is still not clear to me what wn is.

bvallelunga commented on app/encoder.py:
```diff
@@ -0,0 +1,223 @@
+    for idx, x in enumerate(inputs):
+        m = tf.matmul(x, wmx)*tf.matmul(h, wmh)
+        z = tf.matmul(x, wx) + tf.matmul(m, wh) + b
+        i, f, o, u = tf.split(z, 4, 1)
```
Can we have variable names that do not look like they were generated by a compiler?

jkarimi91 replied:
ill look into it; this is the raw code written by the original authors.

bvallelunga commented on app/encoder.py:
```diff
@@ -0,0 +1,223 @@
+        global hps
+        hps = HParams(
+            load_path='model_params/params.jl',
+            nhidden=4096,
```
How are we coming up with these values?

jkarimi91 replied:
Those were the values used to build the model described in the paper this work is presented in.

bvallelunga commented on app/encoder.py:
```diff
@@ -0,0 +1,223 @@
+        def seq_cells(xmb, mmb, smb):
+            return sess.run(cells, {X: xmb, M: mmb, S: smb})
+
+        def transform(xs):
```
Can we clean up this method, it is very long and not sure what it does?

jkarimi91 replied:
I think you are referring to `transform`? ill look into it but at the very least ill add a brief docstring for those not familiar with ML APIs like scikit-l
earn

bvallelunga commented on environment.yml:
```diff
@@ -0,0 +1,10 @@
+name: sentiment
+channels:
+    - default
```
what is channels?

jkarimi91 replied:
you know with pip you can pip install <name> or, in some cases like with tensorflow, you would pip install with a url instead? Well conda (the package manager that uses the `environment.yml` to create a venv) can do the same thing; it can install packages from outside sources as well. To do this, you have to tell conda, in the env file, what sources (aka channels) to look in. Since we are not using any outside sources, we dont even need to list `default` in the `channels` section but i did it to make things easier for development (dont have to google the format etc).

bvallelunga replied:
hmm thats interesting, what are the trade offs of using conda vs pip?

jkarimi91 replied:
Conda > pip in every way:
  * comes with venv creation out of box
  * each venv can even have it's own version of python in it
  * identifies/resolves package dependency conflicts
  * works with pip i.e. if you have a package that needs to be pip installed, you can pip install it within the conda env and even specify packages that should be pip installed by conda, in the `environment.yml` ([example file](https://github.com/jkarimi91/snake/blob/master/environment.yml))
  * it is widely used within the community especially among data scientists, researchers, research/ml engineers and the like; both pytorch and tensorflow list conda install instructions on their install page

Conda is a no brainer. The only time you really see people use pip is when they are first starting out with python and programming. 

An additional note about `channels`, the only time you will really every see a `channel` other than default, is with `pytorch` since facebook hosts torch/pytorch on their own channel at the moment.

bvallelunga committed:
Update README.md

bvallelunga committed:
Update README.md

jay-karimi committed:
Refactoring and removing cruft

The code, written by the original model authors, was very hard to
read, undocumented and generally had bad/inconsistent style.
